arch/rv32i/src/asm.S

```diff
@@ -0,0 +1,5 @@
+.global semihost_command
+semihost_command:
+	slli x0, x0, 0x1f		# 0x01f01013   		Entry NOP
+	.word 0x00100073 		# ebreak	    	Break to debugger
+	.word 0x40705013		# srai x0, x0, 7    NOP encoding the semihosting call number 7
```

Comment:
This confuses me a bit, we use inline asm everywhere else in Tock, and I'm reasonably confident that it's treated as an opaque block of assembly by the compiler. There was some explanation in the top-level PR description, but if this truly cannot be done with inline asm, I think a detailed explanation of why needs to be in this file here.

Reply:
It turns out that the `asm!` blocks are not treated as raw assembly.

The output of this method (linking with a block of code) results in these instructions being in the final binary:

```asm
2000cc64:       01f01013                slli    zero,zero,0x1f
2000cc68:       00100073                ebreak
2000cc6c:       40705013                srai    zero,zero,0x7
```
Which correctly matches the semihosting requirements.

While if instead if I do this in Rust code

```rust
        asm!(
            "slli x0, x0, 0x1f
            ebreak
            srai x0, x0, 7"
        );
```

The final binary contains:

```asm
20009ade:       01f01013                slli    zero,zero,0x1f
20009ae2:       9002                    ebreak
20009ae4:       40705013                srai    zero,zero,0x7
```

As you can see the compiler will compress the `ebreak` instruction, which then doesn't work with the RISC-V semihosting spec. The assembly block on the other hand can use the full hex of the instruction, which avoids any changes to the output.

Reply:
I have added a comment to the assembly file as well

Reply:
Interesting... ARM opcodes resolve the compression of ambiguous instructions with option width specifiers (`mov.n r0 r0 r0` is 16-bit, `mov.w r0 r0 r0` is 32-bit; narrow/wide).

It looks like RISCV sort of does this; the `9002` instruction is more precisely `c.ebreak` (compressed ebreak).

Unfortunately, I can't find an equivalent instruction in RISCV to force the wider instruction choice.

Digging into the RISCV manual, I think the comment as-written is incorrect. If a machine supports compressed instructions, then the assembler is _supposed_ replace `ebreak` with `c.ebreak`. IIUC, the real reason this is working is because the machine model that's passed to the assembler line is _different_ than the true underlying machine model, and claims the architecture doesn't support compression. If I've got this all right, this is pretty subtle and a bit of a (necessary? :/) hack. We should highlight the significance of the "incorrect" machine model given to the assembler in some way though.

It would also be good to verify that there's no way to simply ask the assembler to emit the preferred form of `ebreak`. Following this:
<img width="843" alt="image" src="https://user-images.githubusercontent.com/339422/121939863-796e1280-cd1b-11eb-8afd-99b28d9cd8cb.png">
I tried `g.ebreak`, but that got rejected by the assembler on my machine :/

Reply:
I have updated the comment to try and be more clear that we are disabling compressed instructions as we need the uncompressed version.